design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+defining a single longer array of function pointers, where the first set of function pointers are
+defined by the base interface while the next set of function pointers are defined by the derived
+interface. Multiple levels of COM interface single inheritance is supported.
+
+> Note, while xlang uses COM interfaces as a stable contract between components, xlang is
+explicitly avoiding the rest of the COM standard, including concepts such as apartments, monikers
+and marshalling. These constructs are tightly bound to the Windows platform and so have little use
+in the xlang project. Likewise, xlang is explicitly avoiding Windows technologies that build on COM
+such as DCOM, MTS, COM+, ActiveX and the Windows Runtime. Windows Runtime is an inspiration for
+xlang, but remains a separate implementation.
+
+## IUnknown
+
+> This section is intended to be a short introduction to the IUnknown COM interface.
+For a more thorough explanation of IUnknown, please review the official Microsoft documentation
+[Using and Implementing IUnknown](https://docs.microsoft.com/en-us/windows/desktop/com/using-and-implementing-iunknown).
+
+IUnknown is the root COM interface in any COM-based system. It provides functions for two core
+capabilities - navigating an object's capabilities and managing the object's lifetime. All COM
+interfaces derive from IUnknown and all objects in any COM based system must implement these
+methods.
+
+``` cpp
+struct IUnknown
+{
+    int32_t  QueryInterface(guid const& id, void** object);
+    uint32_t AddRef();
+    uint32_t Release();
+};
+```
+
+### QueryInterface
+
+Given any COM interface pointer, you can query for any other COM interface via IUnknown's
+QueryInterface method. If the referenced COM object supports the interface being queried, a pointer
+to that COM interface is returned to the caller. Otherwise, an error code indicating that the
+requested interface is not supported.
+
+As stated above, interfaces are identified by a GUID. This GUID is passed by reference as the first
+parameter of QueryInterface. The pointer to the requested interface - if available - is returned
+via a void** out parameter. It is the responsibility of the calling code to cast this pointer to
+the correct COM interface pointer type.
+
+Methods on COM interfaces (including xlang ABI interfaces) never throw C++ exceptions, due to the
+need for a stable contract across compilers. QueryInterface returns a 32-bit signed integer as an
+error code known. On Windows, this code is known as an
+[HRESULT](https://docs.microsoft.com/en-us/windows/desktop/com/error-handling-in-com).
+Generally with HRESULTs, negative integers indicate errors, positive integers and zero indicate
+success. QueryInterface in particular should always return 0 on success.
+
+QueryInterface may return the following error codes:
+
+- 0x80004002 (E_NOINTERFACE) indicates the requested interface is not supported by this object.
```

Comment:
I suppose a follow question is, do we ever expect custom QI implementations that may return errors besides success and no interface? If its a true binary state would this return be better as a bool (we don't have RPC and other things that need to intercede here so I'm not sure if we want an "always return an error code" style everywhere)

Reply:
We need to retain IUnknown as-is for broad compatibility and interoperability.

Reply:
hooray baggage :-) Figured this would be the case but want to make sure we are probing our assumptions since we are getting a somewhat "clean slate" here.

Reply:
I'd love a time machine... 😉